src/feature/hs/hs_service.c

```diff
@@ -3042,13 +3042,85 @@ service_desc_hsdirs_changed(const hs_service_t *service,
   return should_reupload;
 }
 
+/** These are all the reasons why a descriptor upload can't occur. We use
+ * those to log the reason properly with the right rate limiting and for the
+ * right descriptor. */
+typedef enum {
+  LOG_DESC_UPLOAD_REASON_MISSING_IPS = 0,
+  LOG_DESC_UPLOAD_REASON_IP_NOT_ESTABLISHED = 1,
+  LOG_DESC_UPLOAD_REASON_NOT_TIME = 2,
+  LOG_DESC_UPLOAD_REASON_NO_LIVE_CONSENSUS = 3,
+  LOG_DESC_UPLOAD_REASON_NO_DIRINFO = 4,
+
+  /* Always +1 of the last value. This is used to allocate the static array of
+   * all rate limiting objects. */
+  LOG_DESC_UPLOAD_REASON_NUM_ = 5,
+} log_desc_upload_reason_t;
+
+/** Log the reason why we can't upload the given descriptor for the given
+ * service. This takes a message string (allocated by the caller) and a
+ * reason.
+ *
+ * Depending on the reason and descriptor, different rate limit applies. This
+ * is done because this function will basically be called every second. Each
+ * descriptor for each reason uses its own log rate limit object in order to
+ * avoid message suppression for different reasons and descriptors. */
+static void
+log_cant_upload_desc(const hs_service_t *service,
+                     const hs_service_descriptor_t *desc, const char *msg,
+                     const log_desc_upload_reason_t reason)
+{
+  static ratelim_t limits[2][LOG_DESC_UPLOAD_REASON_NUM_];
+  bool is_next_desc = false;
+  int rlim_seconds = 0;
+  unsigned int rlim_pos = 0;
+  ratelim_t *rlim = NULL;
+
+  tor_assert(service);
+  tor_assert(desc);
+  tor_assert(msg);
+
+  /* Make sure the reason value is valid. It should never happen because we
+   * control that value in the code flow but will be apparent during
+   * development if a reason is added but LOG_DESC_UPLOAD_REASON_NUM_ is not
+   * updated. */
+  if (BUG(reason > (LOG_DESC_UPLOAD_REASON_NUM_ - 1) || reason < 0)) {
```

Comment:
I'm not sure if I like the `- 1` here. I think it's more standard to do a +1 in the array definition, and then do the check as is. Also if we use sizeof as suggested above we won't fear that the size can change under our feet.

Reply:
Fixed with the previous fix where it is now `MAX` and the +1 is done when we allocate.